### PR TITLE
Breaking change / performance: don't make kubernetes-client deserialize k8s events into objects

### PR DESCRIPTION
**Pull request by rmoe**
Turning API responses into Python objects take a significant
amount of CPU time when dealing with a large number of events.

Fixes: #423

---

## Breaking changes

- The Kubernetes EventsReflector, which is providing the KubeSpawner instances with information about [Kubernetes Events](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#event-v1-core) describing events for other resources, is now exposing events as python dictionaries rather than `V1Event` objects. `V1Event` is defined in the kubernetes-client/python library as a representation of a Kubernetes Event.
- KubeSpawner's `.progress` method implementation (https://github.com/jupyterhub/jupyterhub/pull/1771) which is generating a formatted `message` as well as a KubeSpawner specific `raw_event` entry now returns the `raw_event` as a Python dictionary with entries formatted in `camelCase` where the keys were formatted in `snake_case`.

### CODE REVIEW

File: kubespawner/reflector.py
Comment (mriedem) [inline kubespawner/reflector.py:164]: Technically this changes the value on the `resources` dict from an object to a dict so the help here should probably be updated as well:

https://github.com/jupyterhub/kubespawner/pull/424/files#diff-2f4d4efedf466e7bf4f8f083d0c722adR52

One thing I'm not sure about is how much this might impact out-of-tree reflectors that are expecting `self.resources` to return objects rather than dicts, i.e. this could be a breaking change for them. One could maybe mitigate that by wrapping the dicts in simple objects that pass through `__getattr__` to `__getitem__` but as we can see below with `resource_version` we'd also have to camel-case-ify some of the attributes, like [this](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/fcf9c2e92a450e175a9d53abbe6dbbb2936df096/jupyterhub/files/hub/jupyterhub_config.py#L39).

I'm not sure what the ABI contract on something like this is within the KubeSpawner. Would a release note be sufficient?

File: kubespawner/spawner.py
Comment (mriedem) [inline kubespawner/spawner.py:109]: This is a good example of the kind of issue I'm talking about above. Handling this for the known event and pod reflectors within this repo is sufficient but I worry about breaking out of tree reflectors, but like I said I also don't know what contract is there.

File: kubespawner/spawner.py
Comment (mriedem) [inline kubespawner/spawner.py:1497]: Hmm, maybe this should be:

`not pod["metadata"].get("deletionTimestamp")`

Because `deletionTimestamp` not being in the dict vs being in the dict with a None value are different things, unless that's not possible with how the kube API works?

File: kubespawner/spawner.py
Comment (mriedem) [inline kubespawner/spawner.py:1659]: This comment about the object not being serializable can go away.

File: kubespawner/spawner.py
Comment (mriedem) [inline kubespawner/spawner.py:1649]: nit: unrelated change (try to avoid these because they can cause merge conflicts and bloat the size of the review)

File: kubespawner/spawner.py
Comment (mriedem) [inline kubespawner/spawner.py:1560]: The docstring on the `PodReflector.pods` attribute should probably be updated because it talks about the value being an object representation which it no longer is.

https://github.com/jupyterhub/kubespawner/pull/424/files#diff-3082573191d96b79e2ec71986d4ef3e2R69

**Comment by mriedem** [comment]
>The bigger concern is the impact to out of tree reflectors

By the way, I'm assuming you can have out of tree reflectors but I'm not entirely sure if you can. The KubeSpawner docs don't mention anything about entry points to load your own reflectors:

https://jupyterhub-kubespawner.readthedocs.io/en/latest/reflector.html

File: kubespawner/spawner.py
Comment (rmoe) [reply kubespawner/spawner.py:1497]: `deletionTimestamp` is only set by the server when a deletion is requested. It shouldn't exist at any other time.

File: kubespawner/spawner.py
Comment (rmoe) [reply kubespawner/spawner.py:1560]: Good catch. I'll get all of these comments and docstrings fixed.

File: kubespawner/reflector.py
Comment (consideRatio) [inline kubespawner/reflector.py:161]: Is it correct that introducing partial accomplished the same thing as just adding _preload_content=False to the list of arguments? If so, is there a motivation to using partial over just adding the argument to the list?

```suggestion
        initial_resources = getattr(self.api, self.list_method_name)(
            self.namespace,
            label_selector=self.label_selector,
            field_selector=self.field_selector,
            _request_timeout=self.request_timeout,
            _preload_content=False,
        )
```

**Comment by consideRatio** [comment]
## Notification / Question

@clkao this PR will impact the users of `raw_event` returned by `progress()`. I think the difference will be that you get camelCase instead of snake_case, but still get a python dictionary. @rmoe do you think I described this change correctly?

## My conclusion

Overall, I think this PR makes a lot of sense given the excellent report in #423. I think we should accept it, verify the breaking changes and report them in a release which needs to be minor rather than patch.

## Breaking changes
This is what I pick up to be breaking changes.

- Progress() returned `raw_event` changed from python dictionary formatted with snake_case to a python dictionary formatted with camelCase (I think).
- A spawner's events property populated by the Singleton events reflector now returns python dictionaries instead of V1Event objects from kubernetes-client/python.

**Comment by mriedem** [comment]
FWIW I agree.

>Progress() returned raw_event changed from python dictionary formatted with snake_case to a python dictionary formatted with camelCase (I think).

`raw_event` was a dict because the V1Event object was converted to a dict [here](https://github.com/jupyterhub/kubespawner/pull/424/files#diff-3082573191d96b79e2ec71986d4ef3e2L1659). I'm not sure what `event.to_dict()` does regarding the field names but assuming they were snake-case because that's [what they were](https://github.com/jupyterhub/kubespawner/pull/424/files#diff-3082573191d96b79e2ec71986d4ef3e2L1666) on the `event` object and [now they are camel-case](https://github.com/jupyterhub/kubespawner/pull/424/files#diff-3082573191d96b79e2ec71986d4ef3e2R1660).

>A spawner's events property populated by the Singleton events reflector now returns python dictionaries instead of V1Event objects from kubernetes-client/python.

Yup, the docstring for the `events` property function was updated to reflect that [here](https://github.com/jupyterhub/kubespawner/pull/424/files#diff-3082573191d96b79e2ec71986d4ef3e2R91).

I [mentioned](https://github.com/jupyterhub/kubespawner/pull/424/files#r470732969) that we could try to still return an object with some massaging of the field names but that could get complicated. We'd probably know if that worked if we didn't have to make some of the changes in this patch, but it's hard to say for sure since there are no tests (that I could find) for the EventReflector code.

File: kubespawner/reflector.py
Comment (rmoe) [reply kubespawner/reflector.py:161]: There was no reason. I just copied what I did in `_watch_and_update` (where partial application is required).

**Comment by rmoe** [comment]
> @clkao this PR will impact the users of `raw_event` returned by `progress()`. I think the difference will be that you get camelCase instead of snake_case, but still get a python dictionary. @rmoe do you think I described this change correctly?

That's correct. They were snake case because of how those OpenAPI objects got converted to dictionaries. Now we just have the response from the Kubernetes API as a dictionary and everything the API returns is camel cased.

**Comment by consideRatio** [comment]
@rmoe thank you soo much for your thorough investigation and this PR to resolve the found performance issue! Thank you @mriedem for your review work! :heart: :100: :tada:

@jupyterhub/jupyterhubteam this LGTM but it would be nice to get another LGTM before I press merge. It solves a real problem when scaling to a very large amount of users so it motivates some breaking changes I think.

**Review by consideRatio — APPROVED** [review]
Excellent work! :heart:

**Comment by yuvipanda** [comment]
\o/ Thanks a lot for this, @rmoe and @mriedem! 

And thanks for your review, @consideRatio :)

**Comment by yuvipanda** [comment]
I was *just* dealing with an outage caused by slow response times cascading because of high CPU usage... :)

**Pull request merged**
